### ISSUE TITLE
canvas.line freezes indefinitely with numba 0.51.1

### ISSUE THREAD

**Issue by AjayThorve** [holoviz/datashader#956]
#### ALL software version info
python 3.8
numba 0.51.1
datashader 0.11.1
bokeh 2.2.0
numpy  1.19.1

#### Description of expected behavior and the observed behavior
Observed behavior: `canvas.line` function freezes indefinitely
#### Complete, minimal, self-contained example code that reproduces the issue

```
# code goes here between backticks
import datetime
import pandas as pd
import numpy as np
import xarray as xr
import datashader as ds
import datashader.transfer_functions as tf
from collections import OrderedDict
# Constants
np.random.seed(2)
n = 100000                               # Number of points
cols = list('abcdefg')                   # Column names of samples
start = datetime.datetime(2010, 10, 1, 0)   # Start time
# Generate a fake signal
signal = np.random.normal(0, 0.3, size=n).cumsum() + 50
# Generate many noisy samples from the signal
noise = lambda var, bias, n: np.random.normal(bias, var, n)
data = {c: signal + noise(1, 10*(np.random.random() - 0.5), n) for c in cols}
# Add some "rogue lines" that differ from the rest 
cols += ['x'] ; data['x'] = signal + np.random.normal(0, 0.02, size=n).cumsum() # Gradually diverges
cols += ['y'] ; data['y'] = signal + noise(1, 20*(np.random.random() - 0.5), n) # Much noisier
cols += ['z'] ; data['z'] = signal # No noise at all
# Pick a few samples from the first line and really blow them out
locs = np.random.choice(n, 10)
data['a'][locs] *= 2
# Create a dataframe
data['Time'] = [start + datetime.timedelta(minutes=1)*i for i in range(n)]
df = pd.DataFrame(data)
df['ITime'] = pd.to_datetime(df['Time']).astype('int64')
# Default plot ranges:
x_range = (df.iloc[0].ITime, df.iloc[-1].ITime)
y_range = (1.2*signal.min(), 1.2*signal.max())
print("x_range: {0} y_range: {1}".format(x_range,y_range))
cvs = ds.Canvas(x_range=x_range, y_range=y_range, plot_height=300, plot_width=900)
aggs= OrderedDict((c, cvs.line(df, 'ITime', c)) for c in cols)
img = tf.shade(aggs['a'])
```

cc @philippjfr

**Comment by jakirkham** [comment]
cc @gmarkall

**Comment by quasiben** [comment]
Can you also test with numba 0.50.0 ?

**Comment by AjayThorve** [comment]
@quasiben datashader had regression issues with 0.50.0, the only released version >0.49 that this works is numba 0.51.0

**Comment by jakirkham** [comment]
Interesting. So this worked with 0.51.0 and not 0.51.1?

**Comment by AjayThorve** [comment]
@jakirkham yes! and this time its regressed even on CPU(pandas) version along with cuDF

**Comment by jakirkham** [comment]
Are we able to narrow this down further (like with a Numba only reproducer)? There are only a handful of changes in that patch release. So this should hopefully give us a better idea of what could have caused it.

**Comment by quasiben** [comment]
After digging for a bit it might have something to do with https://github.com/numba/numba/pull/6146

**Comment by philippjfr** [comment]
I can definitely reproduce it. If you interrupt it, it's getting stuck some in here:

    ~/miniconda3/envs/datashader/lib/python3.7/site-packages/numba/core/controlflow.py in _find_back_edges(self)
 
So I also think it's something to do with https://github.com/numba/numba/pull/6146

**Comment by jakirkham** [comment]
cc @sklam (in case you have insights here 🙂)

**Comment by quasiben** [comment]
It does finish but takes some time:

```
real    34m46.416s
user    32m22.290s
sys     0m6.085s
```

**Comment by gmarkall** [comment]
This seems to be specifically related to the change to use a DFS algorithm to find back-edges instead of the dominator-based analysis. If I change back to the old algorithm with:

```diff
diff --git a/numba/core/controlflow.py b/numba/core/controlflow.py
index f68572031..f37e9ae3d 100644
--- a/numba/core/controlflow.py
+++ b/numba/core/controlflow.py
@@ -572,40 +572,14 @@ class CFGraph(object):
         Find back edges.  An edge (src, dest) is a back edge if and
         only if *dest* dominates *src*.
         """
-
-        # Uses a simple DFS to find back-edges.
-        # The new algorithm is faster than the the previous dominator based
-        # algorithm.
         back_edges = set()
-        # stack: keeps track of the traversal path
-        stack = []
-        # succs_state: keep track of unvisited successors of a node
-        succs_state = {}
-        entry_point = self.entry_point()
-
-        def push_state(node):
-            stack.append(node)
-            succs_state[node] = [dest for dest in self._succs[node]]
-
-        push_state(entry_point)

-        while stack:
-            tos = stack[-1]
-            tos_succs = succs_state[tos]
-            # Are there successors not checked?
-            if tos_succs:
-                # Check the next successor
-                cur_node = tos_succs.pop()
-                # Is it in our traversal path?
-                if cur_node in stack:
-                    # Yes, it's a backedge
-                    back_edges.add((tos, cur_node))
-                else:
-                    # Push
-                    push_state(cur_node)
-            else:
-                # Checked all successors. Pop
-                stack.pop()
+        for src, succs in self._succs.items():
+            back = self._doms[src] & succs
+            # In CPython bytecode, at most one back edge can flow from a
+            # given block.
+            assert len(back) <= 1
+            back_edges.update((src, dest) for dest in back)

         return back_edges

```

Then the example completes in a timely fashion again:

```
$ time python repro.py
x_range: (1285891200000000000, 1291891140000000000) y_range: (-104.8094774985483, 100.09816883332905)

real    0m9.732s
user    0m9.601s
sys     0m0.113s
```

**Comment by gmarkall** [comment]
With numba/numba#6186 the execution time goes back down to 10 seconds (what it was for me with 0.51.0) but I don't know that it's doing the right thing because the reproducer produces no output. Is there a way I can check that Datashader is working properly with Numba 0.51.1 + the patch?

**Comment by AjayThorve** [comment]
can confirm this is now fixed with numba 0.51.2, @gmarkall thanks for the quick fix. 



> but I don't know that it's doing the right thing because the reproducer produces no output. Is there a way I can check that Datashader is working properly with Numba 0.51.1 + the patch? the final img output should be something like this (which it now is)
![image](https://user-images.githubusercontent.com/20476096/92151393-e4c2a580-edd5-11ea-851d-8c45f125b280.png)
